AppKit/CPTextView/CPTextView.j

```diff
@@ -0,0 +1,1890 @@
+    [self _fixupReplaceForRange:CPMakeRange(aRange.location, [aString length])];
+}
+
+- (void)insertText:(CPString)aString
+{
+    var isAttributed = [aString isKindOfClass:CPAttributedString],
+        string = isAttributed ? [aString string]:aString;
+
+    if (![self shouldChangeTextInRange:CPMakeRangeCopy(_selectionRange) replacementString:string])
+        return;
+
+    var undoManager = [[self window] undoManager];
+
+    if (isAttributed)
+    {
+        [[undoManager prepareWithInvocationTarget:self]
+                _replaceCharactersInRange:CPMakeRange(_selectionRange.location, [aString length])
+                withAttributedString:[_textStorage attributedSubstringFromRange:CPMakeRangeCopy(_selectionRange)]];
+
+        [undoManager setActionName:@"Replace rich text"];
+        [_textStorage replaceCharactersInRange:CPMakeRangeCopy(_selectionRange) withAttributedString:aString];
+    }
+    else
+    {
+        [undoManager setActionName:@"Replace plain text"];
+
+        if ([self isRichText])
+        {
+            aString = [[CPAttributedString alloc] initWithString:aString attributes:_typingAttributes];
+            [[undoManager prepareWithInvocationTarget:self]
+                _replaceCharactersInRange:CPMakeRange(_selectionRange.location, [aString length])
+                withAttributedString:[_textStorage attributedSubstringFromRange:CPMakeRangeCopy(_selectionRange)]];
+
+            [_textStorage replaceCharactersInRange:CPMakeRangeCopy(_selectionRange) withAttributedString:aString];
+        }
+        else
+        {
+            [[undoManager prepareWithInvocationTarget:self] _replaceCharactersInRange:CPMakeRange(_selectionRange.location, [aString length]) withString:[[self string] substringWithRange:CPMakeRangeCopy(_selectionRange)]];
+            [_textStorage replaceCharactersInRange:CPMakeRangeCopy(_selectionRange) withString:aString];
+        }
+    }
+
+    [self setSelectedRange:CPMakeRange(_selectionRange.location + [string length], 0)];
+    [self didChangeText];
+    [_layoutManager _validateLayoutAndGlyphs];
+    [self sizeToFit];
+    [self scrollRangeToVisible:_selectionRange];
+    _stickyXLocation = _caretRect.origin.x;
+}
+
+- (void)_blinkCaret:(CPTimer)aTimer
+{
+    _drawCaret = !_drawCaret;
+    [self setNeedsDisplayInRect:_caretRect];
+}
+
+
+#pragma mark -
+#pragma mark Drawing methods
+
+- (void)drawInsertionPointInRect:(CGRect)aRect color:(CPColor)aColor turnedOn:(BOOL)flag
+{
+#if PLATFORM(DOM)
+    var style;
+
+    if (!_caretDOM)
+    {
+        _caretDOM = document.createElement("span");
+        style = _caretDOM.style;
+        style.position = "absolute";
+        style.visibility = "visible";
+        style.padding = "0px";
+        style.margin = "0px";
+        style.whiteSpace = "pre";
+        style.backgroundColor = "black";
+        _caretDOM.style.width = "1px";
+        _DOMElement.appendChild(_caretDOM);
```

Comment:
In my opinion we should create a private class _CPCaret. So we can separate the creation of the caret and the textView.